File: backend/app/routers/admin/reviews.py
from fastapi import APIRouter, Depends, HTTPException
from bson import ObjectId
from app.database import get_database
from app.middleware.auth import require_admin
from app.schemas.review import UpdateReviewRequest, ReviewResponse

router = APIRouter()


@router.get("")
async def list_all_reviews(admin: dict = Depends(require_admin)):
    db = get_database()
    cursor = db.reviews.find().sort("created_at", -1)
    items = []
    async for review in cursor:
        items.append(ReviewResponse(
            id=str(review["_id"]),
            author_name=review["author_name"],
            rating=review["rating"],
            body=review["body"],
            is_approved=review["is_approved"],
            created_at=review["created_at"],
        ))
    return {"reviews": items}


@router.put("/{review_id}")
async def update_review(review_id: str, body: UpdateReviewRequest, admin: dict = Depends(require_admin)):
    db = get_database()
    update = {k: v for k, v in body.model_dump(exclude_none=True).items()}
    if not update:
        raise HTTPException(status_code=400, detail="No fields to update")
    result = await db.reviews.update_one({"_id": ObjectId(review_id)}, {"$set": update})
    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail="Review not found")
    return {"message": "Updated"}


@router.delete("/{review_id}")
async def delete_review(review_id: str, admin: dict = Depends(require_admin)):
    db = get_database()
    result = await db.reviews.delete_one({"_id": ObjectId(review_id)})
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="Review not found")
    return {"message": "Deleted"}

File: src/routes/admin/reviews.ts
import { Router, Response } from 'express';
import { ObjectId } from 'mongodb';
import { getDb } from '../../database';
import { requireAdmin, AuthRequest } from '../../middleware/auth';

const router = Router();

router.get('/', requireAdmin, async (_req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const reviews = await db.collection('reviews').find().sort({ created_at: -1 }).toArray();
  res.json({ reviews: reviews.map((r) => ({ id: r._id.toString(), author_name: r.author_name, rating: r.rating, body: r.body, is_approved: r.is_approved, created_at: r.created_at })) });
});

router.put('/:reviewId', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const allowed = ['is_approved', 'body', 'author_name', 'rating'];
  const update: Record<string, unknown> = {};
  for (const key of allowed) {
    if (req.body[key] !== undefined) update[key] = req.body[key];
  }
  if (!Object.keys(update).length) { res.status(400).json({ detail: 'No fields to update' }); return; }
  const db = await getDb();
  const result = await db.collection('reviews').updateOne({ _id: new ObjectId(req.params.reviewId) }, { $set: update });
  if (!result.matchedCount) { res.status(404).json({ detail: 'Review not found' }); return; }
  res.json({ message: 'Updated' });
});

router.delete('/:reviewId', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const result = await db.collection('reviews').deleteOne({ _id: new ObjectId(req.params.reviewId) });
  if (!result.deletedCount) { res.status(404).json({ detail: 'Review not found' }); return; }
  res.json({ message: 'Deleted' });
});

export default router;
